Fix bug 1332552. Treat bsos with ttl:null as never expiring.

mostlygeek opened the pull request:
The 1.5 spec says that undefined and ttl=null should mean the BSO never expires.

With this commit, when ttl:null is encounted it will be changed into 100
years (in seconds). The default BSO TTL is no 100 years as well.

However there now there is a subtle difference between undefined
TTL and ttl=null between insert/update behaviour:

TTL is undefined (not in the JSON)

    PUT .../storage/col/bsoId
     - insert: TTL is set to 100 years
     - update: TTL is *not* updated

TTL=null

    PUT .../storage/col/bsoId
     - insert: TTL is set to 100 years
     - update: TTL is set to 100 years

Refs: 
 - https://github.com/mozilla-services/server-syncstorage/pull/56 - add a functional test to prevent regressions
 - https://bugzilla.mozilla.org/show_bug.cgi?id=1332552

mostlygeek committed:
Fix bug 1332552. Treat bsos with ttl:null as never expiring.

The 1.5 spec says that undefined and ttl=null should mean the BSO never expires.
With this commit, when ttl:null is encounted it will be changed into 100
years (in seconds). The default BSO TTL is no 100 years as well.

However there now there is a subtle difference between undefined
TTL and ttl=null between insert/update behaviour:

TTL is undefined (not in the JSON)

    PUT .../storage/col/bsoId
     - insert: TTL is set to 100 years
     - update: TTL is *not* updated

TTL=null

    PUT .../storage/col/bsoId
     - insert: TTL is set to 100 years
     - update: TTL is set to 100 years

rnewman commented:
The difference between missing and null TTLs concerns me. A long, long time ago Sync used to allow refreshing a TTL, or changing other metadata, by sending a record which was missing fields. We removed that in an effort to simplify the server.

I'd suggest doing exactly what the Python server does here. If it preserves an overridden record's TTL when it is `PUT` with no TTL, do that. If it takes the default unlimited TTL, do that.

rnewman commented on syncstorage/db.go:
```diff
@@ -49,8 +49,9 @@ const (
 	SORT_INDEX
 
-	// Keep BSO for 1 year
-	DEFAULT_BSO_TTL = 365 * 24 * 60 * 60 * 1000
```
Does this mean there are bookmarks, passwords, and keys on a production Go server that will magically disappear in 2018?

If so, that's a critical bug that needs to be fixed.

rnewman commented on web/misc.go:
```diff
@@ -168,12 +168,19 @@ func parseIntoBSO(jsonData json.RawMessage, bso *syncstorage.PutBSOInput) *parse
+
+		// bug 1332552, treat `ttl:null` as never expiring
+		// by using 100 years for it (in seconds)
+		if len(r) == 4 && string(r[:]) == "null" {
```
I am flabberghasted that not only does Go put you in this situation, but it also forces you to do bytewise array comparison in this case.

Please make sure you have tests for invalid JSON, such as:

```
"ttl": nu,
"ttl": "null",
"ttl": [],
```

mostlygeek replied:
Sometimes just dealing with it as a []byte is easier. JSON schemas where `{"x":123}`, `{"x":null}` or `{"x":"123"}` is all valid data from clients makes me sad. It's really difficult to work with it in golang since you can't just unmarshal it into structs. 

When json.Unmarshal tries to parse `nu`, `"null"` or `[]` into an `int` type [it'll fail](https://play.golang.org/p/3QeYyUPDlO). That will bubble up into an HTTP 400 for the client.

rnewman reviewed (changes requested):
Action items:

- Add negative tests for malformed inputs.
- Decide on a recovery plan for iOS users who have lost server-side passwords due to incorrect TTL expiry.
- Decide on an investigation and mitigation plan for desktop, Android, and iOS users who have bookmarks, passwords, and crypto/keys that will expire when they are not meant to.

I'm particularly concerned that in 9 months some users might be broken because their key object silently disappears from the server, rendering their data unrecoverable, or with bookmark trees that become inconsistent due to some records disappearing.

mostlygeek commented:
> The difference between missing and null TTLs concerns me. A long, long time ago Sync used to allow refreshing a TTL, or changing other metadata, by sending a record which was missing fields. We removed that in an effort to simplify the server.

The python server has (test_update_of_ttl_without_sending_data)[https://github.com/mozilla-services/server-syncstorage/blob/master/syncstorage/tests/functional/test_storage.py#L1142-L1166] which seems to ensure that behaviour still exists. 

> I'd suggest doing exactly what the Python server does here. If it preserves an overridden record's TTL when it is PUT with no TTL, do that. If it takes the default unlimited TTL, do that.

+1. I believe the functionality now matches the Python server's for TTL handling.

mostlygeek committed:
Add invalid data tests for web.parseIntoBSO. Fix bug where id=null is ok

mostlygeek commented:
@rnewman thanks for the r. I added some negative tests and fixed a `null` being acceptable for `id`. :man_facepalming: 

Let's take the discussion for fixing the data issues into the bug.

rfk commented:
> > A long, long time ago Sync used to allow refreshing a TTL, or changing other metadata, by
> > sending a record which was missing fields. We removed that in an effort to simplify the server.
> The python server has (test_update_of_ttl_without_sending_data) which seems
> to ensure that behaviour still exists.

Indeed, I believe we resurrected the partial-update functionality for sync1.5, although I don't recall the precise use-case for it.

mostlygeek committed:
update circleci to default to server-syncstorage:1.6.2 to include null test

mostlygeek committed:
Remove usage of PYTHON_TEST_CONTAINER in circle.yml

In practice having the ENV var and the hard coded value means you have
to change it in two places. Better to have it hard coded and update it
with a git record.

mostlygeek commented:
@rnewman if you're good with the code here I'd like to merge / close this PR.

rnewman commented:
SGTM!
